src/NSubstitute/Core/DependencyInjection/NSubstituteDefaultFactory.cs

```diff
@@ -0,0 +1,59 @@
+using NSubstitute.Core.Arguments;
+using NSubstitute.Proxies;
+using NSubstitute.Proxies.CastleDynamicProxy;
+using NSubstitute.Proxies.DelegateProxy;
+using NSubstitute.Routing;
+using NSubstitute.Routing.AutoValues;
+
+namespace NSubstitute.Core.DependencyInjection
+{
+    public static class NSubstituteDefaultFactory
+    {
+        /// <summary>
+        /// The default NSubstitute registrations. Feel free to configure the existing container to customize
+        /// and override NSubstitute parts.
+        /// </summary>
+        public static INSubContainer DefaultContainer { get; } = CreateDefaultContainer();
+
+        public static ISubstitutionContext CreateSubstitutionContext() => DefaultContainer.Resolve<ISubstitutionContext>();
+
+        private static INSubContainer CreateDefaultContainer()
+        {
+            return new NSubContainer()
+                .RegisterSingleton<SequenceNumberGenerator, SequenceNumberGenerator>()
+                .RegisterPerRequest<IThreadLocalContext, ThreadLocalContext>()
+                .RegisterPerRequest<IArgumentSpecificationDequeue>(r =>
+                    new ArgumentSpecificationDequeue(r.Resolve<IThreadLocalContext>().DequeueAllArgumentSpecifications))
+                .RegisterPerRequest<ICallSpecificationFactory, CallSpecificationFactory>()
+                .RegisterPerRequest<IArgumentSpecificationFactory, ArgumentSpecificationFactory>()
```

Comment:
It might work, but I'd prefer to do that in a separate PR. It would be much easier to see whether it adds to maintainability, or makes it worse. Given the amount of mess we have here, it will be too much.

WDYT? If you are fine, I'll create a follow up issue.